### PR TITLE
Fix DIP1000 diagnostics for `return` and add tests for both `return` and `scope`

### PR DESCRIPTION
**Pull request by JinShil**
I could not find a diagnostic test for when `scope` was used as a function attribute on functions without a `this` context pointer, but the code was there.

Also, `return` was not emitting a similar error for the same reason as `scope`.  I found that the compiler was actually removing the `return` attribute instead of emitting an error.

This PR fixes both.

### CODE REVIEW

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -3403,7 +3403,13 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
         if (f.isscope && !funcdecl.isNested() && !ad)
         {
-            funcdecl.error("functions cannot be `scope`");
+            /* Non-static nested functions have a hidden 'this' pointer to which
+             * the 'scope' applies
+             */
+            if (sc.scopesym && sc.scopesym.isAggregateDeclaration())
+                funcdecl.error("`static` member has no `this` to which `scope` can apply");
+            else
+                error(funcdecl.loc, "Top-level function `%s` has no `this` to which `scope` can apply", funcdecl.toChars());
```
Comment (JinShil) [inline src/dmd/dsymbolsem.d]: This was modified to emit a similar error message `return` a few lines below.  Please expand the source code to see.

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -1231,11 +1231,6 @@ extern(C++) Type typeSemantic(Type t, Loc loc, Scope* sc)
             }
             if (tf.next.hasWild())
                 wildreturn = true;
-
-            if (tf.isreturn && !tf.isref && !tf.next.hasPointers())
```
Comment (Geod24) [inline src/dmd/typesem.d]: Why this change ?

Comment (JinShil) [reply src/dmd/typesem.d]: Per the 2nd paragraph in this PR's fist comment.

> Also, `return` was not emitting a similar error for the same reason as `scope`. I found that the compiler was actually removing the `return` attribute instead of emitting an error.

The compiler is currently silently removing the`return` attribute instead of emitting an error stating that the `return` attribute is not valid for functions. The compiler already emits an error for `scope` so this change makes `return` behave the same.

File: test/fail_compilation/diag_dip1000.d [hunk 3]
```diff
@@ -0,0 +1,26 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/diag_dip1000.d(13): Error: Top-level function `testScope` has no `this` to which `scope` can apply
+fail_compilation/diag_dip1000.d(14): Error: Top-level function `testReturn` has no `this` to which `return` can apply
```
Comment (Geod24) [inline test/fail_compilation/diag_dip1000.d]: Sorry to nitpick, but that's the only error message with uppercase at the beginning of the message.

**Review by Geod24 — CHANGES_REQUESTED** [review]
Can you add a test for templated function too ?

**Comment by JinShil** [comment]
> Can you add a test for templated function too?

I'm working on it.  Please consider this a WIP until I announce it as ready.

File: src/dmd/dsymbolsem.d [hunk 4]
```diff
@@ -6499,3 +6486,40 @@ void aliasSemantic(AliasDeclaration ds, Scope* sc)
...
+/**
+Checks if a function declaration's lifetime attributes (e.g. `scope` and `return`)
+are valid
+Params:
+    fd = The function declaration to check
+    sc = The semantic scope in which to check
+*/
+private void checkLifetimeAttributes(FuncDeclaration fd, Scope* sc)
+{
+    if (fd.type)
```
Comment (Geod24) [inline src/dmd/dsymbolsem.d]: Code is usually better if you return early. So instead of nesting the whole function in an `if`, use:
```d
if (fd.type)
    return;
```

**Review by Geod24 — APPROVED** [review]
One style nitpick, oLGTM

**Comment by JinShil** [comment]
This PR now checks template functions for the validity of the DIP1000 lifetime attributes `scope` and `return`, but be forewarned that this is the first code to do so.  Currently the compiler does not check function templates for the validity of attributes.  For example.

```D
// Currently...
static abstract void foo(){}    // Non-template: Error: `static` functions cannot be `abstract`
static abstract void bar()(){}  // Template:     No error

// Due to this PR
void abc() scope {}   // Non-template:  Error: top-level function `abc` has no `this` to which `scope` can apply
void def()() scope {} // Template:      Error: top-level function `def` has no `this` to which `scope` can apply
```
https://run.dlang.io/is/e8gxGZ

Creating a holistic solution that validates attributes on template functions would be a rather large endeavor.

**Comment by JinShil** [comment]
Btw, this is ready to go unless you want me to rever the template stuff to keep the status quo.
